src/snowflake/snowpark/_internal/data_source/datasource_reader.py

```diff
@@ -85,6 +88,11 @@ def read(self, partition: str) -> Iterator[List[Any]]:
                         batch = []
             else:
                 raise ValueError("fetch size cannot be smaller than 0")
+        except Exception as exc:
+            if self.driver.non_retryable_error_checker(exc):
+                raise _SnowparkDataSourceNonRetryableException(exc)
```

Comment:
as per our discussion, we do not need the private class and can just raise the existing reader exception here